### PR TITLE
Add datetime length patterns as skeleta and fix non-Gregorian datagen

### PR DESCRIPTION
**Pull request by sffc**
Part of #1317

This PR adds new skeleta `DayComponents::Auto`, `DayComponents::AutoWeekday`, and `TimeComponents::Auto` with data populated from the datetime length patterns.

This PR enables us to centralize on one single code path for datetime formatting (and two public types, one for AnyCalendar and one for concrete calendar), which will be really nice.

It also fixes non-Gregorian neo skeleta datagen.

### CODE REVIEW

**Comment by sffc** [comment]
Note: the size increase from 218376 to 308744 is more due to the bug that was preventing non-Gregorian calendars being in the data than it is about the new skeleton key. The size goes to ___ after adding the new skeleton key but before fixing the bug.

File: components/datetime/src/neo_skeleton.rs [hunk 1]
```diff
@@ -280,8 +302,11 @@ impl NeoDayComponents {
         Self::YearMonthDayWeekday,
         Self::EraYearMonthDayWeekday,
         Self::Weekday,
+        Self::Auto,
+        Self::AutoWeekday,
     ];
 
+    // Note:
```
Comment (Manishearth) [inline components/datetime/src/neo_skeleton.rs:309]: nit: missing?

File: provider/datagen/src/transform/cldr/datetime/mod.rs [hunk 2]
```diff
@@ -51,6 +52,146 @@ fn supported_cals() -> &'static HashMap<icu_locid::extensions::unicode::Value, &
     })
 }
 
+impl DatagenProvider {
+    fn get_datetime_resources(
```
Comment (Manishearth) [inline provider/datagen/src/transform/cldr/datetime/mod.rs:58]: praise: nice

File: components/datetime/src/skeleton/helpers.rs [hunk 3]
```diff
@@ -564,26 +565,38 @@ impl DateTimeFormatterOptions {
                     &components_bag,
                     false,
                 ) {
-                    BestSkeleton::AllFieldsMatch(p) => Some(p),
-                    _ => None,
+                    BestSkeleton::AllFieldsMatch(p) => p,
+                    _ => {
+                        // Build a last-resort pattern that contains all of the requested fields.
+                        // This is NOT in the CLDR standard! Should be using Append Items.
```
Comment (Manishearth) [inline components/datetime/src/skeleton/helpers.rs:571]: question: should we file and link to an issue for this?

**Review by Manishearth — DISMISSED** [review]

File: components/datetime/src/skeleton/helpers.rs (hunk 3 above)
Comment (sffc) [reply components/datetime/src/skeleton/helpers.rs]: Added reference to https://github.com/unicode-org/icu4x/issues/586

File: components/datetime/src/neo_skeleton.rs (hunk 1 above)
Comment (sffc) [reply components/datetime/src/neo_skeleton.rs]: Fixed

**Pull request merged**